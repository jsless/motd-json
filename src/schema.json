{
  "$id": "https://jsless.org/motd.schema.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Messages of the Day",
  "type": "array",
  "minItems": 1,
  "items": {
    "$ref": "#/definitions/motd"
  },
  "definitions": {
    "motd": {
      "type": "object",
      "required": [
        "message"
      ],
      "properties": {
        "message": {
          "type": "string",
          "description": "The message of the day",
          "minLength": 5
        },
        "period": {
          "$ref": "#/definitions/period"
        },
        "regions": {
          "type": "array",
          "minItems": 1,
          "items": {
            "$ref": "#/definitions/region"
          }
        },
        "tags": {
          "type": "array",
          "minItems": 1,
          "items": {
            "oneOf": [
              { "$ref": "#/definitions/booleanTag" },
              { "$ref": "#/definitions/numberTag" },
              { "$ref": "#/definitions/semverTag" },
              { "$ref": "#/definitions/setTag" }
            ]
          }
        }
      }
    },
    "period": {
      "type": "object",
      "required": [
        "from",
        "till"
      ],
      "properties": {
        "from": {
          "type": "string",
          "format": "date"
        },
        "till": {
          "type": "string",
          "format": "date"
        }
      }
    },
    "region": {
      "type": "string"
    },
    "semverTag": {
      "type": "object",
      "required": [
        "name",
        "type",
        "value"
      ],
      "properties": {
        "name": {
          "type": "string"
        },
        "type": {
          "enum": ["semver"]
        },
        "value": {
          "type": "string"
        }
      }
    },
    "numberTag": {
      "type": "object",
      "required": [
        "name",
        "type",
        "value"
      ],
      "properties": {
        "name": {
          "type": "string"
        },
        "type": {
          "enum": ["number"]
        },
        "value": {
          "type": "number"
        }
      }
    },
    "setTag": {
      "type": "object",
      "required": [
        "name",
        "type",
        "value"
      ],
      "properties": {
        "name": {
          "type": "string"
        },
        "type": {
          "enum": ["set"]
        },
        "value": {
          "type": "array"
        }
      }
    },
    "booleanTag": {
      "type": "object",
      "required": [
        "name",
        "type",
        "value"
      ],
      "properties": {
        "name": {
          "type": "string"
        },
        "type": {
          "enum": ["boolean"]
        },
        "value": {
          "type": "boolean"
        }
      }
    }
  }
}

