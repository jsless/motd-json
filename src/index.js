import osLocale from 'os-locale'
import semver from 'semver'
import schema from '../src/schema'

export function getRegion() {
  return osLocale.sync({ spawn: false })
    .substr(3, 2)
    .toLowerCase()
}

export function filterPeriod(data, now) {
  if (now && typeof now === 'string') {
    now = new Date(now)
  }

  now = now || new Date().getTime()
  const currentYear = `${new Date().getFullYear()}-`

  return data.filter(({ period: { from, till } = {} }) => {
    if (from && till) {
      from = new Date(from.replace('0000-', currentYear)).getTime()
      till = new Date(till.replace('0000-', currentYear)).getTime()

      return now >= from && now <= till
    }

    return true
  })
}

export function filterRegional(data, region) {
  if (region) {
    region = region.toLowerCase()
  } else {
    region = getRegion()
  }

  return data.filter(({ regions }) => !regions || regions.includes(region))
}

export function filterTags(data, tags) {
  if (!tags) {
    return data
  }

  return data.filter(({ tags: dataTags }) => {
    if (!dataTags) {
      return true
    }

    return dataTags.every(({ name, type, value }) => {
      if (!(name in tags)) {
        return true
      }

      const tagValue = tags[name]

      if (type === 'semver') {
        return semver.satisfies(tagValue, value)
      }

      if (type === 'number') {
        // TODO: make operator optionsurable
        return tagValue >= value
      }

      if (type === 'set') {
        if (Array.isArray(tagValue)) {
          return tagValue.some(aTagValue => value.includes(aTagValue))
        }
        return value.includes(tagValue)
      }

      if (type === 'boolean') {
        return value === tagValue
      }
    })
  })
}

export function filter(data, options) {
  const {
    region = '',
    regional = false,
    date = '',
    tags = false
  } = options || {}

  if (regional || region) {
    data = filterRegional(data, region)
  }

  if (date) {
    data = filterPeriod(data, date)
  }

  if (tags) {
    data = filterTags(data, tags)
  }

  return data
}

let ajvValidate

export async function validate(data) {
  if (!ajvValidate) {
    try {
      const Ajv = await import('ajv').then(m => m.default || m)
      ajvValidate = new Ajv().compile(schema)
    } catch (err) {
      if (err.code === 'MODULE_NOT_FOUND') {
        console.error('ajv not found, please install the peer dependency') // eslint-disable-line no-console
        return true
      }

      throw err
    }
  }

  return ajvValidate(data)
}

export function motd(data, options) {
  const motds = filter(data, options)

  const index = Math.floor(Math.random() * motds.length)

  const { message } = motds[index]
  return message || ''
}
