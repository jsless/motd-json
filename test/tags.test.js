import * as motd from '../src'

const data = [
  {
    message: 'You are using v2.x',
    tags: [
      {
        name: 'version',
        type: 'semver',
        value: 'v2.x'
      }
    ]
  },
  {
    message: 'You are using v3.x',
    tags: [
      {
        name: 'version',
        type: 'semver',
        value: 'v3.x'
      }
    ]
  },
  {
    message: 'You are using 10 modules or more',
    tags: [
      {
        name: 'modulesCount',
        type: 'number',
        value: 10
      }
    ]
  },
  {
    message: 'You are using the axios or http module',
    tags: [
      {
        name: 'modules',
        type: 'set',
        value: ['axios', 'http']
      }
    ]
  },
  {
    message: 'You are using typescript',
    tags: [
      {
        name: 'typescript',
        type: 'boolean',
        value: true
      }
    ]
  }
]

describe('motd', () => {
  test('schema', async () => {
    await expect(motd.validate(data)).resolves.toBe(true)
  })

  test('filterTags doesnt filter without tags specified', () => {
    const filteredMotds = motd.filterTags(data)

    expect(filteredMotds.length).toBe(5)
    expect(filteredMotds).toMatchSnapshot()
  })

  test('filterTags doesnt filter when data has no tags', () => {
    const data = [
      { message: 'my test message' }
    ]
    const filteredMotds = motd.filterTags(data, {
      version: 'v2.x'
    })

    expect(filteredMotds.length).toBe(1)
    expect(filteredMotds).toMatchSnapshot()
  })

  test('filterTags filters everything', () => {
    const filteredMotds = motd.filterTags(data, {
      version: 'v1.x',
      modulesCount: 1,
      modules: '',
      typescript: false
    })

    expect(filteredMotds.length).toBe(0)
  })

  test('filterTags filters semver', () => {
    const filteredMotds = motd.filterTags(data, {
      version: 'v2.0.1'
    })

    expect(filteredMotds).toEqual(expect.arrayContaining([
      expect.objectContaining({
        message: 'You are using v2.x'
      })
    ]))

    expect(filteredMotds).toEqual(expect.not.arrayContaining([
      expect.objectContaining({
        message: 'You are using v3.x'
      })
    ]))
  })

  test('filterTags filters semver', () => {
    const filteredMotds = motd.filterTags(data, {
      version: 'v3.0.1'
    })

    expect(filteredMotds).toEqual(expect.not.arrayContaining([
      expect.objectContaining({
        message: 'You are using v2.x'
      })
    ]))

    expect(filteredMotds).toEqual(expect.arrayContaining([
      expect.objectContaining({
        message: 'You are using v3.x'
      })
    ]))
  })

  test('filterTags filters boolean', () => {
    const filteredMotds = motd.filterTags(data, {
      typescript: true
    })

    expect(filteredMotds).toEqual(expect.arrayContaining([
      expect.objectContaining({
        message: 'You are using typescript'
      })
    ]))
  })

  test('filterTags filters set', () => {
    const filteredMotds = motd.filterTags(data, {
      modules: ['axios', 'i18n']
    })

    expect(filteredMotds).toEqual(expect.arrayContaining([
      expect.objectContaining({
        message: 'You are using the axios or http module'
      })
    ]))
  })
})
