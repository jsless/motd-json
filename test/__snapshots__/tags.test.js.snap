// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`motd filterTags doesnt filter when data has no tags 1`] = `
Array [
  Object {
    "message": "my test message",
  },
]
`;

exports[`motd filterTags doesnt filter without tags specified 1`] = `
Array [
  Object {
    "message": "You are using v2.x",
    "tags": Array [
      Object {
        "name": "version",
        "type": "semver",
        "value": "v2.x",
      },
    ],
  },
  Object {
    "message": "You are using v3.x",
    "tags": Array [
      Object {
        "name": "version",
        "type": "semver",
        "value": "v3.x",
      },
    ],
  },
  Object {
    "message": "You are using 10 modules or more",
    "tags": Array [
      Object {
        "name": "modulesCount",
        "type": "number",
        "value": 10,
      },
    ],
  },
  Object {
    "message": "You are using the axios or http module",
    "tags": Array [
      Object {
        "name": "modules",
        "type": "set",
        "value": Array [
          "axios",
          "http",
        ],
      },
    ],
  },
  Object {
    "message": "You are using typescript",
    "tags": Array [
      Object {
        "name": "typescript",
        "type": "boolean",
        "value": true,
      },
    ],
  },
]
`;
